File: src/main/kotlin/sunsetsatellite/lang/sunlite/Token.kt
package sunsetsatellite.lang.sunlite

class Token(val type: TokenType, val lexeme: String, val literal: Any?, val line: Int, val file: String?, val pos: Position) {
	override fun toString(): String {
		return "$type $lexeme ${literal ?: ""}"
	}

	data class Position(val start: Int, val end: Int)

	companion object {
		fun unknown(): Token {
			return Token(TokenType.IDENTIFIER, "<unknown>", null, -1, null, Position(-1, -1))
		}

		fun identifier(name: String, line: Int = -1, file: String? = null ): Token {
			return Token(TokenType.IDENTIFIER, name, null, line, file, Position(-1, -1))
		}

		fun identifier(name: String, expr: Expr): Token {
			return Token(TokenType.IDENTIFIER, name, null, expr.getLine(), expr.getFile(), Position(-1, -1))
		}

		fun identifier(name: String, token: Token): Token {
			return Token(TokenType.IDENTIFIER, name, null, token.line, token.file, Position(-1, -1))
		}
	}

}

File: src/main/csharp/SunliteSharp/Core/Token.cs
using SunliteSharp.Core.Enum;

namespace SunliteSharp.Core;

public readonly record struct Token(TokenType Type, string Lexeme, object? Literal, int Line, string File, Token.Position Pos)
{
    public readonly record struct Position(int Start, int End);

    public static Token Identifier(string name, string file)
    {
        return new Token(TokenType.Identifier, name, null, -1, file, new Position(-1, -1));
    }
    
    public static Token Identifier(string name, int line, string file)
    {
        return new Token(TokenType.Identifier, name, null, line, file, new Position(-1, -1));
    }
}